engine/function/src/templates/Numeric.ftl

```diff
@@ -1722,127 +1786,241 @@ public class Numeric {
+                final ${pt.primitive} v = vi.${pt.iteratorNext}();
+
                 if (isNull(result[i - 1])) {
                     result[i] = v;
                 } else if (isNull(v)) {
                     result[i] = result[i - 1];
                 } else {
-                    result[i] = (${pt.primitive}) (result[i - 1] + v);
+                    result[i] = (long) (result[i - 1] + v);
                 }
-    
+
                 i++;
             }
         }
 
         return result;
     }
+    </#if>
 
     /**
      * Returns the cumulative product.  Null values are excluded.
      *
      * @param values values.
      * @return cumulative product of non-null values.
      */
-    public static ${pt.primitive}[] cumprod(${pt.boxed}[] values) {
+    <#if pt.valueType.isFloat >
+    public static double[] cumprod(${pt.boxed}[] values) {
         return cumprod(unbox(values));
     }
+    <#else>
+    public static long[] cumprod(${pt.boxed}[] values) {
+        return cumprod(unbox(values));
+    }
+    </#if>
 
     /**
      * Returns the cumulative product.  Null values are excluded.
      *
      * @param values values.
      * @return cumulative product of non-null values.
      */
-    public static ${pt.primitive}[] cumprod(${pt.primitive}... values) {
+    <#if pt.valueType.isFloat >
+    public static double[] cumprod(${pt.primitive}... values) {
         if (values == null) {
             return null;
         }
 
         return cumprod(new ${pt.vectorDirect}(values));
     }
+    <#else>
+    public static long[] cumprod(${pt.primitive}... values) {
+        if (values == null) {
+            return null;
+        }
+
+        return cumprod(new ${pt.vectorDirect}(values));
+    }
+    </#if>
 
     /**
      * Returns the cumulative product.  Null values are excluded.
      *
      * @param values values.
      * @return cumulative product of non-null values.
      */
-    public static ${pt.primitive}[] cumprod(${pt.vector} values) {
+    <#if pt.valueType.isFloat >
+    public static double[] cumprod(${pt.vector} values) {
         if (values == null) {
             return null;
         }
 
         if (values.isEmpty()) {
-            return new ${pt.primitive}[0];
+            return new double[0];
         }
 
         final int n = values.intSize("cumprod");
-        ${pt.primitive}[] result = new ${pt.primitive}[n];
+        double[] result = new double[n];
 
         try ( final ${pt.vectorIterator} vi = values.iterator() ) {
-            result[0] = vi.${pt.iteratorNext}();
+            final ${pt.primitive} v0 = vi.${pt.iteratorNext}();
+            result[0] = isNull(v0) ? NULL_DOUBLE : v0;
             int i = 1;
     
             while (vi.hasNext()) {
                 final ${pt.primitive} v = vi.${pt.iteratorNext}();
+
+                if (isNaN(v)) {
+                    Arrays.fill(result, i, n, Double.NaN);
+                    return result;
+                } else if (isNull(result[i - 1])) {
```

Comment:
Maybe missing infinity handling?

Reply:
Better short-circuiting was added.  Infinity handling should be fine.

Reply:
Unit test added.